exchanges/stream/websocket.go

```diff
@@ -786,163 +776,134 @@ func (w *Websocket) GetName() string {
+// Errors are returned for duplicates or exceeding max Subscriptions
+func (w *Websocket) SubscribeToChannels(subs subscription.List) error {
+	if slices.Contains(subs, nil) {
+		return common.ErrNilPointer
 	}
-	return nil
-}
-
-// AddSubscription adds a subscription to the subscription lists
-// Unlike AddSubscriptions this method will error if the subscription already exists
-func (w *Websocket) AddSubscription(c *subscription.Subscription) error {
-	w.subscriptionMutex.Lock()
-	defer w.subscriptionMutex.Unlock()
-	if w.subscriptions == nil {
-		w.subscriptions = subscriptionMap{}
+	if err := w.checkSubscriptions(subs); err != nil {
+		return err
 	}
-	key := c.EnsureKeyed()
-	if _, ok := w.subscriptions[key]; ok {
-		return ErrSubscribedAlready
+	if err := w.Subscriber(subs); err != nil {
+		return fmt.Errorf("%w: %w", ErrSubscriptionFailure, err)
 	}
-
-	n := *c // Fresh copy; we don't want to use the pointer we were given and allow encapsulation/locks to be bypassed
-	w.subscriptions[key] = &n
-
 	return nil
 }
 
-// SetSubscriptionState sets an existing subscription state
-// returns an error if the subscription is not found, or the new state is already set
-func (w *Websocket) SetSubscriptionState(c *subscription.Subscription, state subscription.State) error {
-	w.subscriptionMutex.Lock()
-	defer w.subscriptionMutex.Unlock()
-	if w.subscriptions == nil {
-		w.subscriptions = subscriptionMap{}
-	}
-	key := c.EnsureKeyed()
-	p, ok := w.subscriptions[key]
-	if !ok {
-		return ErrSubscriptionNotFound
+// AddSubscriptions adds subscriptions to the subscription store
+// Sets state to Subscribing unless the state is already set
+func (w *Websocket) AddSubscriptions(subs ...*subscription.Subscription) error {
+	if w == nil {
+		return common.ErrNilPointer
 	}
-	if state == p.State {
-		return ErrChannelInStateAlready
+	if w.subscriptions == nil {
+		w.subscriptions = subscription.NewStore()
 	}
-	if state > subscription.UnsubscribingState {
-		return errInvalidChannelState
+	var errs error
+	for _, s := range subs {
+		if s.State() == subscription.InactiveState {
+			if err := s.SetState(subscription.SubscribingState); err != nil {
+				errs = common.AppendError(errs, fmt.Errorf("%w: %s", err, s))
+			}
+		}
+		if err := w.subscriptions.Add(s); err != nil {
+			errs = common.AppendError(errs, err)
+		}
 	}
-	p.State = state
-	return nil
+	return errs
 }
 
-// AddSuccessfulSubscriptions adds subscriptions to the subscription lists that
-// has been successfully subscribed
-func (w *Websocket) AddSuccessfulSubscriptions(channels ...subscription.Subscription) {
-	w.subscriptionMutex.Lock()
-	defer w.subscriptionMutex.Unlock()
+// AddSuccessfulSubscriptions marks subscriptions as subscribed and adds them to the subscription store
+func (w *Websocket) AddSuccessfulSubscriptions(subs ...*subscription.Subscription) error {
+	if w == nil {
+		return common.ErrNilPointer
+	}
 	if w.subscriptions == nil {
```

Comment:
This looks like an error case

Reply:
Same [as above](https://github.com/thrasher-corp/gocryptotrader/pull/1501#discussion_r1623999234), this was deliberate.